sql/src/test/java/io/crate/operation/projectors/FetchProjectorTest.java

```diff
@@ -0,0 +1,190 @@
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.*;
+import java.util.concurrent.TimeUnit;
+
+import static org.hamcrest.core.Is.is;
+import static org.mockito.Mockito.*;
+
+public class FetchProjectorTest extends CrateUnitTest {
+
+    private static final TableIdent USER_TABLE_IDENT = new TableIdent(Schemas.DEFAULT_SCHEMA_NAME, "users");
+    private ThreadPool threadPool;
+
+    @Before
+    public void before() throws Exception {
+        threadPool = new ThreadPool("testThreadPool");
+    }
+
+    @After
+    public void after() throws Exception {
+        threadPool.shutdown();
+        threadPool.awaitTermination(1, TimeUnit.SECONDS);
+    }
+
+    @Test
+    public void testFetchSize() throws Throwable {
+        CollectingRowReceiver rowReceiver = new CollectingRowReceiver();
+
+        TransportFetchOperation transportFetchOperation = mock(TransportFetchOperation.class);
+        SettableFuture<IntObjectMap<? extends Bucket>> future = SettableFuture.create();
+        when(transportFetchOperation.fetch(anyString(), any(IntObjectMap.class), anyBoolean()))
+            .thenReturn(future);
+        RowUpstream upstream = mock(RowUpstream.class);
+
+        int fetchSize = 3;
+        FetchProjector fetchProjector = prepareFetchProjector(fetchSize, rowReceiver, transportFetchOperation, upstream);
+
+        IntObjectMap<Bucket> intObjectMap = new IntObjectHashMap<>(2);
+        Collection<Object[]> rows = new ArrayList<>(10);
+
+        int i;
+        for (i = 1; i <= 10; i++) {
+            Row row = new Row1((long)i);
+            rows.add(new Object[]{row});
+            CollectionBucket collectionBucket = new CollectionBucket(rows);
+            intObjectMap.put(0 , collectionBucket);
+
+            fetchProjector.setNextRow(row);
+            if (i % fetchSize == 0) {
+                verify(upstream, atLeastOnce()).pause();
+                future.set(intObjectMap);
+                verify(upstream, timeout(100).atLeastOnce()).resume(false);
+            }
+        }
+        assertThat(i, is(11));
+        verify(upstream, timeout(100).times(3)).pause();
+        verify(upstream, timeout(100).times(3)).resume(false);
+        fetchProjector.finish();
+
+        Bucket projected = rowReceiver.result();
+        assertThat(projected.size(), is(10));
+
+        int iterateLength = Iterables.size(rowReceiver.result());
+        assertThat(iterateLength, is(10));
+    }
+
+
+    private FetchProjector prepareFetchProjector(int fetchSize,
+                                                 CollectingRowReceiver rowReceiver,
+                                                 TransportFetchOperation transportFetchOperation,
+                                                 RowUpstream upstream) {
+        FetchProjector pipe =
+            new FetchProjector(
+                transportFetchOperation,
+                threadPool.executor(ThreadPool.Names.SUGGEST),
```

Comment:
You could use `Executors.new...` here instead of creating a ThreadPool (ThreadPool creates a dozen of threads which is a waste for the test)